File: chap51.py

n = 3
arr = [0, 5, 9]

subset = []
def search(k):
    if k == n:
        print subset
    else:
        search(k + 1)
        subset.append(arr[k])
        search(k + 1)
        del subset[-1]

search(0)

import itertools

print list(itertools.permutations([1,3,9]))

File: chap51.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> subset;
int n;

void search(int k) {
    if (k == n) {
        for (auto i : subset)
            cout << i << " ";
        cout << "\n";
    }
    else {
        search(k + 1);
        subset.push_back(k);
        search(k + 1);
        subset.pop_back();
    }
}

int main() {
    n = 4;
    search(0);
    return 0;
}